### PR TITLE
python_requirement: Check for development headers

### PR DESCRIPTION
**Pull request by bw-yuhogg**
Note that the mere presence of \<include\>/python2.7 is not enough
to prove the Python headers are installed and that there are at
least two different versions of python-config slithering around,
one of which (Debian's?) has a --configdir option and one of which
does not.

Signed-off-by: Bob W. Hogg <rwhogg@linux.com>

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

The idea here is to require use of a brewed Python if the system Python lacks either the shared library or the development headers.

This is an early draft; a couple of things I should note:

* I'm not 100% sure my logic is right; my knowledge of Python configuration is far from perfect.
* I have yet to test this with Python 3
* I have only tested this on Ubuntu 17.04. I'll want to test it on at least Debian and CentOS before merging.
* I have not run `brew tests`; I'll worry about that later

I'm doing this on Linuxbrew first so we can iron out the kinks here; that will make porting to upstream easier IMO.

### CODE REVIEW

**Comment by sjackman** [comment]
Looks good to me. Two comments. Can you factor out `devel_installed?` from `which_python`? If `Python.h` is installed, on most distributions I think we can assume that the `.so` library is also installed, since these files are installed by the `lib*-dev` or `-devel` package. So I think we could simplify the check and look only for the header file. You've already written the code to check for the shared library too, so I'm fine with keeping it. For the equivalent `:ruby` code check out https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/requirements/ruby_requirement.rb#L41

**Comment by sjackman** [comment]
Thanks, Bob! This fix will eliminate a number of support requests from people with a Python interpreter installed without the `-dev` package, which is most people I think.

**Comment by maxim-belkin** [comment]
I somehow removed my comments - sorry about that.

The main points were:
- use `sys.prefix` (python2) and `sys.base_prefix` (python3) to detect prefix (instead of pythonX-config)
- use something like `directory_of(python_executable)` instead of plain `python_executable` (not sure if this is necessary

**Comment by bw-yuhogg** [comment]
>  So I think we could simplify the check and look only for the header file. You've already written the code to check for the shared library too, so I'm fine with keeping it.

I'd rather just remove it. We'll get better performance and having the headers but not a shared library makes no sense - what use would they be without it*?

\* Arguably they'd still be useful with a static library version of Python, but that's another matter entirely...

File: Library/Homebrew/requirements/python_requirement.rb
Comment (sjackman) [inline Library/Homebrew/requirements/python_requirement.rb]: `or` is usually `||` for style?
```ruby
return python_executable if OS.mac? || devel_installed?(python_executable)
```

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by sjackman** [comment]
Two minor style changes. Looks good to me otherwise. Thanks, Bob!

File: Library/Homebrew/requirements/python_requirement.rb
Comment (sjackman) [inline Library/Homebrew/requirements/python_requirement.rb]: `brew style` says this `return` is unnecessary.

**Comment by bw-yuhogg** [comment]
Thanks for the review Shaun! Unfortunately, there's one final kink I discovered: the include paths for Python 3 now have an "m" appended. (Why, I do not know) Debian fixes this by symlinking /usr/include/python3.6 to /usr/include/python3.6m, but Homebrew does not.

I'm working on a fix for this, should be ready shortly.

(EDIT: fixed)

**Comment by sjackman** [comment]
Here's the explanation regarding `m`: https://stackoverflow.com/questions/16675865/difference-between-python3-and-python3m-executables

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
Can you please wait until I merge 1.2.2 before merging this PR? Hopefully tomorrow I'll have 1.2.2 merged. https://github.com/Linuxbrew/brew/pull/410

**Comment by sjackman** [comment]
I've merged `brew` 1.2.2. You can merge this PR now.

**Comment by sjackman** [comment]
Thanks, Bob!

**Pull request merged**
